### ISSUE TITLE
Remote write stop sending samples

### ISSUE THREAD

**Issue by xenofree** [prometheus/prometheus#10264]
**What did you see ? Under which circumstances?**

Prometheus stop sending samples to remote write.
I see in log that desired shards exploded

ts=2022-02-07T21:25:56.883Z caller=dedupe.go:112 component=remote level=info remote_name=29a4b7 url=http://x.x.x.x:yyy/insert/0/prometheus/api/v1/write msg="Remote storage resharding" from=1 to=2

ts=2022-02-07T21:26:06.883Z caller=dedupe.go:112 component=remote level=info remote_name=29a4b7 url=http://x.x.x.x:yyy/insert/0/prometheus/api/v1/write msg="Remote storage resharding" from=2 to=3

ts=2022-02-07T21:26:16.884Z caller=dedupe.go:112 component=remote level=info remote_name=29a4b7 url=http://x.x.x.x:yyy/insert/0/prometheus/api/v1/write msg="Remote storage resharding" from=3 to=5

...

ts=2022-02-07T21:27:26.883Z caller=dedupe.go:112 component=remote level=debug remote_name=29a4b7 url=http://x.x.x.x:yyy/insert/0/prometheus/api/v1/write msg=QueueManager.updateShardsLoop lowerBound=30.79999
9999999997 desiredShards=53.58719996305498 upperBound=57.2

ts=2022-02-07T21:39:16.884Z caller=dedupe.go:112 component=remote level=warn remote_name=29a4b7 url=http://x.x.x.x:yyy/insert/0/prometheus/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644269277 minSendTimestamp=1644269946

![prometheus stats](https://user-images.githubusercontent.com/35098906/152952664-cbf61255-9ea8-4afc-8dca-c74eeff5c9d9.PNG)

**Environment**

Prometheus docker version.
prometheus, version 2.33.1 (branch: HEAD, revision: 4e08110891fd5177f9174c4179bc38d789985a13)
  build user:       root@37fc1ebac798
  build date:       20220202-15:23:18
  go version:       go1.17.6
  platform:         linux/amd64

* Prometheus configuration file:
```
global:
  scrape_interval: 1m
  scrape_timeout: 55s
  evaluation_interval: 30s
  external_labels:
    datacenter: xxxxx

remote_write:
  - url: http://x.x.x.x:yyyy/insert/0/prometheus/api/v1/write
    queue_config:
      batch_send_deadline: 5s
      min_backoff: 1s
      max_backoff: 30s
```

**Comment by roidelapluie** [comment]
Are you using Prometheus Agent?

**Comment by xenofree** [comment]
No

services:
  prometheus:
    image: prom/prometheus:v2.33.1
    container_name: prometheus
    restart: always
    ports:
      - "9090:9090"
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/etc/prometheus/tsdb'
      - '--storage.tsdb.retention.size=35GB'
      - '--web.enable-lifecycle'
      - '--log.level=debug'

**Comment by roidelapluie** [comment]
Thank you. It would be helpful to get a goroutine dump. you  can fetch it at http://prometheus:9090/debug/pprof/goroutine?debug=2

**Comment by xenofree** [comment]
I have the problem again.

i dumped you the goroutine

[goroutine_dump_prometheus_remote_write_lag.txt](https://github.com/prometheus/prometheus/files/8024164/goroutine_dump_prometheus_remote_write_lag.txt)

**Comment by roidelapluie** [comment]
Are you remote writing with HTTPS?

**Comment by xenofree** [comment]
No, i'm writting with http

```
remote_write:
  - url: http://x.x.x.x:yyyy/insert/0/prometheus/api/v1/write
    queue_config:
      batch_send_deadline: 5s
      min_backoff: 1s
      max_backoff: 30s
```

**Comment by hagen1778** [comment]
Hi @xenofree! I think, limiting the number of [max shards](https://prometheus.io/docs/practices/remote_write/#max_shards) is a sensible thing to do in this case. It would keep the number of opened connections to the remote storage predictable.

**Comment by xenofree** [comment]
@hagen1778 Thanks for the tip. Is there a best practice for calculating max shards ? Like maybe 1 shard per cpu ?

**Comment by hagen1778** [comment]
@xenofree yes, that is a good estimation.
From the URL format you use, I assume the remote storage is VictoriaMetrics. If it is, please follow [these instructions](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#prometheus-setup) for remote write configuration. 
My apologies if the assumption is not correct.

**Comment by roidelapluie** [comment]
cc @cstyan @csmarchbanks @bwplotka   could you have a look at this issue to help finding what could cause this? Thanks!

**Comment by fsperling** [comment]
We had the same problem twice since we upgraded to v2.33.1.
We have a remote write to a VictoriaMetricsDB in AWS.
Once it happened when we upgraded the EKS cluster VictoriaMetrics lives on which caused a short unavailability of VictoriaMetrics.

**Comment by fsperling** [comment]
@bboreham you changed a lot in the remote_write functionality since 2.31.x
Could you please have a look too.

**Comment by bboreham** [comment]
The symptoms are consistent with your backend stopping.

> Is there a best practice for calculating max shards ? 

You need to know the rate you want to send samples, the typical round-trip time to the backend, and number of samples per send (which you can configure).

Suppose you have 1 million series being scraped every 15 seconds, that is 67K samples/sec.
Suppose your RTT is 100ms, then you can do 10 sends per second.  If you send 500 samples per send (the default), then you get 5,000 samples/sec per shard.  So in this case you want ~14 shards for steady-state. 
You could double that to allow for variability and catch-up after an outage.  Maybe 50 as an extreme max.

**Comment by steve-exley** [comment]
Thanks for all the useful info.

We have seen these symptoms repeatedly. Version 2.33.1

the logs show repeated "Skipping resharding, last successful send was beyond threshold", messages.  Once we get this, remote_write seems to stop indefinitely.

This is where they start:

```
ts=2022-02-10T01:02:42.106Z caller=dedupe.go:112 component=remote level=warn remote_name=10937f url=http://prometheus-collector:9090/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644454950 minSendTimestamp=1644454952

ts=2022-02-10T01:02:24.472Z caller=dedupe.go:112 component=remote level=info remote_name=10937f url=http://prometheus-collector:9090/api/v1/write msg="Currently resharding, skipping."

ts=2022-02-10T01:02:12.311Z caller=dedupe.go:112 component=remote level=info remote_name=10937f url=http://prometheus-collector:9090/api/v1/write msg="Remote storage resharding" from=1 to=2
```
I've included the debug dump as previous posts.
[prometheus.log](https://github.com/prometheus/prometheus/files/8037428/prometheus.log)


I can't find a metric that shows the remote write rtt per request to do the math on the number of shards.

We will increase the min_shards for now to see if that avoids the problem.

**Comment by csmarchbanks** [comment]
Hello, just starting to look at this. @xenofree in your goroutine dump it looks like there is one shard that is blocked on sending data (see the goroutine starting at line 435), then a few other shards that cannot receive data anymore as the buffer has filled. That sounds like there is some sort of issue with the remote system, or writing to the remote system. What do `rate(prometheus_remote_storage_samples_retried_total[5m])` and `rate(prometheus_remote_storage_samples_total[5m])` look like during the period of time where storage is backed up? That would give use an idea if a shard is just completely blocked for some reason or if it is at least trying to send the samples. Resharding does appear to be trying to stop a queue in your dump, I am investigating why that appears to be blocked.

@steve-exley you can see request duration from remote write by looking at the `prometheus_remote_storage_sent_batch_duration_seconds` histogram. E.g. `rate(prometheus_remote_storage_sent_batch_duration_seconds_count[5m])` for request rate and `histogram_quantile(0.95, rate(prometheus_remote_storage_sent_batch_duration_seconds_bucket[5m]))` for latency. I am not seeing any indication from your goroutine dump that resharding is occurring at the time of the dump.

Are there any log messages with `msg="Failed to send batch, retrying"` for either of you? Also, if you happen to have tracing set up for Prometheus you might be able to find a slow/failing trace.

**Comment by csmarchbanks** [comment]
I think I found where a deadlock is happening when resharding and one of the shards hits its batch_send_deadline. I am going to try to get a PR with a fix out today. In the meantime you can set min_shards = max_shards at a reasonably high value to make sure that code path is not hit.

**Comment by steve-exley** [comment]
Thank @csmarchbanks ,

I have built a docker image with the fix and deployed it to one of our clusters. The issue has been occurring intermittently so it will be hard to confirm it is fixed.

I looked for the `prometheus_remote_storage_sent_batch_duration_seconds ` and other metrics on the receiving Prometheus instance and got no result. The metrics may be on the source Prometheus instances, but I cannot query those in agent mode.

I could not find any occurrences of `msg="Failed to send batch, retrying"` in the logs.

I did review occurrences of `msg="Remote storage resharding"` and can see this appears infrequently and normally there are no adverse side effects.

If it happens again on the new code, I'll report back.

**Comment by csmarchbanks** [comment]
[2.33.2](https://github.com/prometheus/prometheus/releases/tag/v2.33.2) and [2.33.3](https://github.com/prometheus/prometheus/releases/tag/v2.33.3) have been released with what I believe to be a fix for this. Let me know if you still see this issue and I will dig in further.

**Comment by steve-exley** [comment]
Thanks @csmarchbanks . No issues so far with the patched version. We have deployed the new release over multiple clusters now and will continue to monitor.

**Comment by reduakt** [comment]
We didn't have any problems in prior versions, but now we found this occurs since February 15 (version 2.33.3): 
```
Feb 15 22:00:10 prometheus prometheus[4332]: ts=2022-02-15T21:00:10.632Z caller=compact.go:519 level=info component=tsdb msg="write block" mint=1644948000875 maxt=1644955200000 ulid=01FVZJYQE9CNX34V91N02XS7XX duration=9.279073947s
Feb 15 22:00:10 prometheus prometheus[4332]: ts=2022-02-15T21:00:10.674Z caller=db.go:1287 level=info component=tsdb msg="Deleting obsolete block" block=01FVT79J690X0MKDNK41NZ15D7
Feb 15 22:00:11 prometheus prometheus[4332]: ts=2022-02-15T21:00:11.619Z caller=head.go:818 level=info component=tsdb msg="Head GC completed" duration=944.876104ms
Feb 15 23:36:52 prometheus prometheus[4332]: ts=2022-02-15T22:36:52.879Z caller=dedupe.go:112 component=remote level=warn remote_name=5390db url=http://127.0.0.1:8425/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644963501 minSendTimestamp=1644964602
Feb 15 23:41:02 prometheus prometheus[4332]: ts=2022-02-15T22:41:02.878Z caller=dedupe.go:112 component=remote level=warn remote_name=5390db url=http://127.0.0.1:8425/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644963501 minSendTimestamp=1644964852
Feb 15 23:41:52 prometheus prometheus[4332]: ts=2022-02-15T22:41:52.879Z caller=dedupe.go:112 component=remote level=warn remote_name=5390db url=http://127.0.0.1:8425/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644963501 minSendTimestamp=1644964902
Feb 15 23:42:02 prometheus prometheus[4332]: ts=2022-02-15T22:42:02.879Z caller=dedupe.go:112 component=remote level=warn remote_name=5390db url=http://127.0.0.1:8425/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644963501 minSendTimestamp=1644964912
```
```
prometheus, version 2.33.3 (branch: HEAD, revision: 56e14463bccfbb6a8facfb663fed5e0ca9f8b387)
  build user:       root@4ee34e4f7340
  build date:       20220211-20:48:21
  go version:       go1.17.7
  platform:         linux/amd64
```
Does anyone have any idea why this is happening? The fix should be applied, but the problem can still occur?

**Comment by csmarchbanks** [comment]
Hmm, according to your logs resharding did not occur in the first place so this may be a different issue. The deadlock that I fixed was only possible to encounter during reshard or shutdown. Unless your log level is set to warn instead of info? Also, would you be able to provide a goroutine dump?

Just to verify, you are also seeing that sending never starts again? Some of the results of the metrics mentioned in https://github.com/prometheus/prometheus/issues/10264#issuecomment-1035027922 could also be helpful.

**Comment by baryluk** [comment]
We are also experiencing this issue. We configure remote_write just few days ago, and after initial one day of testing how it behaves when network connection is disrupted, were happy, then deployed it.

After network routing issues, after network was restored, it had issues (including some Out-of-order samples), but then even hours later, the sender says:

```
Feb 15 23:42:02 prometheus prometheus[4332]: ts=2022-02-15T22:42:02.879Z caller=dedupe.go:112 component=remote level=warn remote_name=5390db url=http://127.0.0.1:8425/api/v1/write msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1644963501 minSendTimestamp=1644964912
```

every 10 seconds.

We are using 2.33.4 on the receiving side, but 2.33.1 on the sender side. I just upgraded to 2.33.4 on a sender side, so we will see if it help.

**Comment by csmarchbanks** [comment]
The issue I fixed was on the sender side, so definitely post again if you see it after upgrading to 2.33.4, ideally with a goroutine dump so I can debug further!

**Comment by hilman-k8n** [comment]
facing the same issue w/ kube-prometheus-stack:32.2.1

This happen usually after 2 days prometheus server running
Upgraded to 33.0.0. Let's see if the issue still persist

**Comment by EmmaTinten** [comment]
Hey there,

I've got a logfile with a dump of the goroutines. I'll attach the corrsponding file.

The service stopped writing this morning with 2.33.4.

Thanks!
[prometheus_log.zip](https://github.com/prometheus/prometheus/files/8177873/prometheus_log.zip)

**Comment by csmarchbanks** [comment]
Thank you for the goroutine dump! I will investigate today and report back what I find/open a PR.

**Comment by csmarchbanks** [comment]
This PR should fix the issue I saw in your goroutine dump: https://github.com/prometheus/prometheus/pull/10395. I don't know yet if we will release another patch for v2.33 or if it will go into v2.34.

**Comment by EmmaTinten** [comment]
@csmarchbanks 

I thank you for investigating!

Here's the corresponding remote write config: 

```
remote_write:
- url: http://127.0.0.1:8428/api/v1/write
  queue_config:
    max_samples_per_send: 500
    max_shards: 30
- url: http://127.0.0.1:8425/api/v1/write
  queue_config:
    max_samples_per_send: 500
    max_shards: 30
```

**Comment by csmarchbanks** [comment]
Thanks! That means max_samples_per_send is 4x larger than capacity, and the fix I have will avoid the deadlock. I would also recommend increasing capacity to something larger than max_samples_per_send, the general guideline is ~5x. That allows slow requests in one shard to avoid blocking all of the other shards. This will also let you avoid the issue until the bugfix is released.

**Comment by EmmaTinten** [comment]
Yeah, right. The capacity itself has never been touched before but the problem hits us first with 2.33.3. 
The corresponding system will be reconfigured to match the capacity within the next days.

Thank you!

**Comment by steve-exley** [comment]
[go-dump.log](https://github.com/prometheus/prometheus/files/8193748/go-dump.log)

Thanks for persisting with this. We've had less issues since increasing the min_shards, but it did occur again yesterday.
Version: 2.33.3

Go dump attached:  
```
remote_write:
- url: http://[host]:9508/api/v1/write
  remote_timeout: 30s
  follow_redirects: true
  queue_config:
    capacity: 2500
    max_shards: 200
    min_shards: 8
    max_samples_per_send: 500
    batch_send_deadline: 5s
    min_backoff: 30ms
    max_backoff: 5s
```

**Comment by csmarchbanks** [comment]
Thank you for the report @steve-exley, that looks to be the same deadlock that @EmmaTinten ran into, though your capacity is larger than your max_samples_per_send, so it should be less likely to happen. Perhaps the path I was worried about in https://github.com/prometheus/prometheus/pull/10395#discussion_r819702464 happened. It is certainly possible with large targets such as kube state metrics. Feel free to run the code in that PR if you can and report back if you see another. If not there should be another patch release this week.

**Comment by csmarchbanks** [comment]
With the release of 2.33.5 I am hoping this can be closed, but will leave it open for another week or two in case of more reports!

**Comment by steve-exley** [comment]
We are running 2.33.5, no issues yet. I will report back if that changes.

**Comment by baryluk** [comment]
After 2 weeks on our side, also no issues.

**Comment by EmmaTinten** [comment]
We have this new version running since the day of publishing without problems. It looks good. Thanks again!

**Comment by csmarchbanks** [comment]
Yay! Glad to hear that the new version is doing well. I am going to close this but I am happy to reopen if an issue does occur for anyone.

**Comment by eugenyuk** [comment]
Hello @csmarchbanks,

We have very similar issue with our prometheus:v2.33.5.
This is what is in a log:
`ts=2022-04-16T15:00:16.222Z caller=compact.go:519 level=info component=tsdb msg="write block" mint=1650110400000 maxt=1650117600000 ulid=01G0SE6MG4EJ4FSPGMYFHNDAHD duration=16.089291156s
ts=2022-04-16T15:00:16.241Z caller=db.go:1287 level=info component=tsdb msg="Deleting obsolete block" block=01G0PMY67YFJ60HFCXQ81CMWGR
ts=2022-04-16T15:00:16.862Z caller=head.go:827 level=info component=tsdb msg="Head GC completed" duration=620.481959ms
ts=2022-04-16T15:00:16.892Z caller=checkpoint.go:98 level=info component=tsdb msg="Creating checkpoint" from_segment=24332 to_segment=24335 mint=1650117600000
ts=2022-04-16T15:00:22.845Z caller=head.go:996 level=info component=tsdb msg="WAL checkpoint complete" first=24332 last=24335 duration=5.953561862s
ts=2022-04-16T15:02:28.938Z caller=dedupe.go:112 component=remote level=info remote_name=936c33 url=http://vm-victoria-metrics-cluster-vminsert:8480/insert/0/prometheus msg="Remote storage resharding" f
rom=1 to=2
ts=2022-04-16T15:02:44.120Z caller=dedupe.go:112 component=remote level=warn remote_name=936c33 url=http://vm-victoria-metrics-cluster-vminsert:8480/insert/0/prometheus msg="Skipping resharding, last su
ccessful send was beyond threshold" lastSendTimestamp=1650121343 minSendTimestamp=1650121354
ts=2022-04-16T15:02:54.124Z caller=dedupe.go:112 component=remote level=warn remote_name=936c33 url=http://vm-victoria-metrics-cluster-vminsert:8480/insert/0/prometheus msg="Skipping resharding, last successful send was beyond threshold" lastSendTimestamp=1650121343 minSendTimestamp=1650121364
`

This is remote_write configuration:
`remote_write:
- url: http://vm-victoria-metrics-cluster-vminsert:8480/insert/0/prometheus
  remote_timeout: 30s`
 
Goroutine dump is in attachment.

Please assist.
[prometheus-goroutine-debug.out.txt](https://github.com/prometheus/prometheus/files/8512566/prometheus-goroutine-debug.out.txt)

**Comment by csmarchbanks** [comment]
Thank you for the report and goroutine dump, I have opened https://github.com/prometheus/prometheus/pull/10608 which should fix the issue you see. Hopefully it will be released with v2.35.0 very shortly.

**Comment by joaoluchetti** [comment]
Hi,
Getting this same issue on Prometheus 2.36.2:

`ts=2022-09-14T14:26:44.995Z caller=main.go:957 level=info msg="Server is ready to receive web requests."
ts=2022-09-14T14:26:45.005Z caller=manager.go:937 level=info component="rule manager" msg="Starting rule manager..."
ts=2022-09-14T14:27:17.809Z caller=dedupe.go:112 component=remote level=info remote_name=79a964 url=http://arquivei-mimir.dev-arquivei.com.br/api/v1/push msg="Done replaying WAL" duration=33.196455981s
ts=2022-09-14T14:27:24.613Z caller=dedupe.go:112 component=remote level=info remote_name=79a964 url=http://arquivei-mimir.dev-arquivei.com.br/api/v1/push msg="Remote storage resharding" from=1 to=6
ts=2022-09-14T14:27:44.613Z caller=dedupe.go:112 component=remote level=info remote_name=79a964 url=http://arquivei-mimir.dev-arquivei.com.br/api/v1/push msg="Remote storage resharding" from=6 to=3
`
After that Prometheus keep restarting over and over again, any ideas on how to fix?

**Comment by bboreham** [comment]
@joaoluchetti do you know the reason for the restarts?  Is that logged somewhere, e.g. by whatever is restarting it?

**Comment by joaoluchetti** [comment]
@bboreham The reason for the restart is the readines probes start failing after a while: 

`Warning  Unhealthy               2m26s (x145 over 16m)  kubelet                                Readiness probe failed: HTTP probe failed with statuscode: 503`

Regarding Prometheus-server logs the latest entries that it shows are like the ones i sent above:

```ts=2022-09-14T15:26:13.476Z caller=dedupe.go:112 component=remote level=info remote_name=840851 url=http://arquivei-mimir.dev-arquivei.com.br/api/v1/push msg="Done replaying WAL" duration=23.282529628s
ts=2022-09-14T15:26:26.039Z caller=dedupe.go:112 component=remote level=info remote_name=840851 url=http://arquivei-mimir.dev-arquivei.com.br/api/v1/push msg="Remote storage resharding" from=8 to=13
ts=2022-09-14T15:26:33.206Z caller=dedupe.go:112 component=remote level=info remote_name=840851 url=http://arquivei-mimir.dev-arquivei.com.br/api/v1/push msg="Remote storage resharding" from=13 to=17
```

**Comment by csmarchbanks** [comment]
Hmm, the readiness probe failures don't sound like this issue. This issue specifically caused remote-write to stop sending but the rest of Prometheus continued to work well. To be sure is it possible to get a goroutine dump?

Are you using the `/-/ready` endpoint for the readiness check? As far as I know the only times it returns a 503 would be during startup or after receiving a termination request. It is possible something is blocking in the configuration load, which would also be revealed with a goroutine dump.

**Comment by bboreham** [comment]
Also, readiness probes do not cause a restart.
Look in your kubelet logs at the time Prometheus was killed; it should say something relevant.

**Comment by joaoluchetti** [comment]
@csmarchbanks, **goroutine.txt** dump is attached
[goroutine.txt](https://github.com/prometheus/prometheus/files/9569481/goroutine.txt)

Also, i am using standard liveness and readiness probes, i didn't change probes configuration.
Forgot to mention but livenessProbe is also failing after a while running Prometheus-server:
``` 
 Warning  Unhealthy               72s                kubelet                                Liveness probe failed: Get "http://10.40.155.16:9090/-/healthy": dial tcp 10.40.155.16:9090: connect: connection refused
  Warning  Unhealthy               62s (x4 over 72s)  kubelet                                Readiness probe failed: Get "http://10.40.155.16:9090/-/ready": dial tcp 10.40.155.16:9090: connect: connection refused
  Normal   Pulled                  60s                kubelet                                Container image "quay.io/prometheus/prometheus:v2.36.2" already present on machine
  Normal   Created                 45s (x2 over 12m)  kubelet                                Created container prometheus-server
  Normal   Started                 44s (x2 over 12m)  kubelet                                Started container prometheus-server
  Warning  Unhealthy               2s (x3 over 12s)   kubelet                                Readiness probe failed: HTTP probe failed with statuscode: 503
```
@bboreham, regarding Kubelet logs, i currently don't have access to them, will try to get those (need to enable cloud logging first since we are using GKE cluster) and return here.

**Comment by joaoluchetti** [comment]
Hi Guys,

It turns out the problem was related to memory, by activating remoteWrite the memory usage increased a little bit. The node where the pod was running didn't have enough resources, so i created a new node with more memory and now Prometheus is running good for almost 2 hours.

Thanks a lot for the suggestions and help!

**Comment by sklarsa** [comment]
Hello,

We're experiencing this problem running Prometheus v2.37 (upgraded to v2.40.7 but that didn't help).  I made a change last week that removed around 3/4 of the metrics that we are writing with remote write to telegraf.  We now seeing that intermittently, remote write just stops.  Eventually, after an hour or so, it catches back up.

I've reviewed https://grafana.com/blog/2021/04/12/how-to-troubleshoot-remote-write-issues-in-prometheus/, but we're not sending a high volume of metrics, so I haven't changed any of the shard settings.

Here's a sample of highest timestamp vs queue highest sent timestamp.  Also attaching goroutine dump:
![Screen Shot 2022-12-19 at 3 00 21 PM](https://user-images.githubusercontent.com/1929541/208509845-1499099a-1816-4e8d-8a43-db541d23dbf3.png)

[goroutine-dump.txt](https://github.com/prometheus/prometheus/files/10262378/goroutine-dump.txt)

I've checked metrics and we're fine on memory usage; below the resource limit.

Thanks for your help!

**Comment by bboreham** [comment]
@sklarsa what was the time of that goroutine dump?

It doesn't indicate anything stuck, for instance this would say something like "3 minutes" if it had been waiting that long:
```
goroutine 485 [select]:
github.com/prometheus/prometheus/storage/remote.(*shards).runShard(0xc008af3170, {0x3bc7cb8, 0xc000797f80}, 0x24fda4a?, 0xc0006cdea0)
	/app/storage/remote/queue_manager.go:1326 +0x347
created by github.com/prometheus/prometheus/storage/remote.(*shards).start
	/app/storage/remote/queue_manager.go:1073 +0x3b3
```

Ideally we would like a dump at 12:50 or 15:00, in the timescale of the metrics chart you provided.

**Comment by sklarsa** [comment]
@bboreham we were actually able to reproduce this issue.  When I removed metrics that we were remote writing, I only kept a set of specific custom metrics that we care about.   These metrics aren't present for every pod in our cluster, only a few.  What happened was, I removed all pods with the custom metric, so nothing was being sent over `remote_write`, which is why the last written timestamp was constant... because there literally was nothing to write! Eventually, this would trigger an alert.  Once I added a new pod with the custom metric, `remote_write` started working as expected.  So I guess everything is working just fine :-)

**Comment by JayChanggithub** [comment]
I'm meet same issue in Prometheus revision `v2.42.0`. Have any exactly solutions for those?  `prometheus-prometheus-0 prometheus ts=2023-04-20T07:58:19.562Z caller=dedupe.go:112 component=remote level=warn remote_name=d405ca url=http://thanos-receive.xxxxxxxxxxx/api/v1/receive msg="Failed to send batch, retrying" err="Post \"http://thanos-receive.thanos.eurekacloud.io/api/v1/receive\": context deadline exceeded"`

**Comment by LeoHsiao1** [comment]
I've been using prometheus v2.42 in remote write mode for a few months and it suddenly started reporting the following error today:
```sh
{"caller":"dedupe.go:112","component":"remote","count":9951,"err":"server returned HTTP status 400 Bad Request: out of bounds","exemplarCount":0,"level":"error","msg":"non-recoverable error","remote_name":"prometheus-main","ts":"2023-05-20T06:30:39.640Z","url":"https://prometheus.test.com/api/v1/write"}
{"caller":"dedupe.go:112","component":"remote","err":"Post \"https://prometheus.ops.skieer.com/api/v1/write\": context deadline exceeded","level":"warn","msg":"Failed to send batch, retrying","remote_name":"prometheus-main","ts":"2023-05-20T06:31:08.049Z","url":"https://prometheus.test.com/api/v1/write"}
```

The nginx logs show that some remote write requests succeed and some fail. The time consumption of these http requests is also different.
![image](https://github.com/prometheus/prometheus/assets/45813403/6dab68dd-3304-4269-b5ba-1a6577dc277c)

I guess the reason is http request timeout. So I increased the threshold:
```yml
remote_write:
- url: [http://[host]:9508/api/v1/write](https://prometheus.test.com/api/v1/write)
  remote_timeout: 60s    # The default value is 30s
```

Continuing to look at the logs, I noticed a decrease in error reporting.
Further, why does the http request time out? I tested the network quality with the ping command. I found that my prometheus in the US, writing data to my prometheus in China, increased the network packet loss rate from the usual 1% to 10%.
Therefore, I enabled the `bbr` congestion control algorithm for the linux server. Then prometheus no longer reports errors.

**Comment by bboreham** [comment]
Please open a new issue. 
This issue is closed and your error message is different.
